### ISSUE TITLE
Slow Customer Login & Checkout (Related to Quote Issues)

### ISSUE THREAD

**Issue by christopher-silvey** [magento/magento2#24391]
### Preconditions (*)

v2.3.2
AWS 4 core 32GB Memory (Plenty of resources)

### Steps to reproduce (*)

I'm not 100% sure how to reproduce this but I do know it happens when something is wrong with the user's quotes. Maybe it is missing?

### Expected result (*)

Expect to login quickly and checkout quickly.

### Actual result (*)

Nothing but a spinning wheel that eats up your CPU and Memory resources and eventually causes the site owner to restart httpd or the page just eventually times out. I fixed this issue on more than one occasion by simply truncating the quotes and then the login is fast. Since I am in development I never have more than 10 quote items in the table when this happens so it doesn't have anything to do with the size of the quote tables. 

When this happens again, and it will, I'll try to see what the actual query is to pinpoint the exact quote issue and report back if I am able to do so.

**Comment by christopher-silvey** [comment]
I actually just found this which shows it is happening to others and backs up what I thought was the cause :)

https://magento.stackexchange.com/questions/284158/some-magento2-customers-login-takes-too-much-time#answer-284218

**Comment by christopher-silvey** [comment]
It is happening again. See image. It happened yesterday to the point all the CPU was being utilized (I use AWS xlarge EC2 instance with plenty of CPU and memory) and customers could not log in. I finally had to delete the rows in the quote table and that fixed the problem yet again.

<img width="1489" alt="Screen Shot 2019-09-05 at 7 52 22 AM" src="https://user-images.githubusercontent.com/3162422/64343654-88565b80-cfb2-11e9-82ea-b47678099ec3.png">

That was after hitting the clear cart button with the user being logged in already. The cart did not clear either.

When this happens CPU usage spikes to above 75% or more and after clearing the quote table it drops down to an overage of 0.02%. Hopefully that provides some insight to how bad this issue can get.

What I can tell you so far is that customer address ID indeed exists correctly in the customer address table and in the quote table. Not sure what is going on but I'm about to debug this and provide more feedback when available.

**Comment by christopher-silvey** [comment]
Ok, I found the issue. In the Magento\Quote\Model\QuoteAddressValidator class the following function was the issue (119 - 131):

```
    /**
     * Validate address to be used for cart.
     *
     * @param CartInterface $cart
     * @param AddressInterface $address
     * @return void
     * @throws \Magento\Framework\Exception\InputException The specified address belongs to another customer.
     * @throws \Magento\Framework\Exception\NoSuchEntityException The specified customer ID or address ID is not valid.
     */
    public function validateForCart(CartInterface $cart, AddressInterface $address): void
    {
        $this->doValidate($address, $cart->getCustomerIsGuest() ? null : $cart->getCustomer()->getId());
    }
```

In my case the logged in user was still returning true for `$cart->getCustomerIsGuest()` which returned null for customer ID. I fixed this by doing the following:

```
    /**
     * Validate address to be used for cart.
     *
     * @param CartInterface $cart
     * @param AddressInterface $address
     * @return void
     * @throws \Magento\Framework\Exception\InputException The specified address belongs to another customer.
     * @throws \Magento\Framework\Exception\NoSuchEntityException The specified customer ID or address ID is not valid.
     */
    public function validateForCart(CartInterface $cart, AddressInterface $address): void
    {
        $this->doValidate($address, $cart->getCustomer()->getId() ? $cart->getCustomer()->getId() : false);
    }
```

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-20324` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by christopher-silvey** [comment]
No problem. I created the fix here: Major Quote Bugfix #24685

Thanks!

**Comment by magento-engcom-team** [comment]
Hi @magentoman2009. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-20324` was closed as `non-reproducible` in `2.3-develop`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean Magento instance or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.
